npm-toolset tests fail with new release of ubi8 Node.js image

Red Hat published new releases of a lot of ubi8 images (https://access.redhat.com/errata/RHBA-2023:0132) on 2023-01-12 which includes `ubi8/nodejs-16` and `ubi8/nodejs-18`.

This causes our tests which invoke `jest` via `npx` to fail with the following error:

```
> src@1.0.0 test
> JEST_JUNIT_OUTPUT_DIR='build/test-results/test' JEST_JUNIT_OUTPUT_NAME='report.xml' npx jest --reporters=default --reporters=jest-junit --coverage --coverageDirectory=build/coverage --forceExit

npm ERR! code ENOENT
npm ERR! syscall lstat
npm ERR! path /opt/app-root/src/.npm-global
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, lstat '/opt/app-root/src/.npm-global'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent
```

Using the previous release of `ubi8/nodejs-16` via `FROM registry.access.redhat.com/ubi8/nodejs-16:1-72` does not produce this error.

However it seems that all previous releases of `ubi8/nodejs-18` produce this error, but further investigation is needed.

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -17,7 +17,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        image: ["package-image", "finish", "go-toolset", "gradle-toolset", "helm", "sonar", "start", "pipeline-manager", "python-toolset", "node16-npm-toolset"]
+        image: ["package-image", "finish", "go-toolset", "gradle-toolset", "helm", "sonar", "start", "pipeline-manager", "python-toolset", "node16-npm-toolset", "node18-npm-toolset"]
     steps:
       -
         name: Checkout
@@ -53,7 +53,7 @@ jobs:
     runs-on: ubuntu-latest
     needs: build-images
     env:
-      IMAGES: package-image finish go-toolset gradle-toolset helm sonar start pipeline-manager python-toolset node16-npm-toolset
+      IMAGES: package-image finish go-toolset gradle-toolset helm sonar start pipeline-manager python-toolset node16-npm-toolset node18-npm-toolset
     steps:
       -
         name: Download image artifacts

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -30,7 +30,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        image: ["package-image", "finish", "go-toolset", "gradle-toolset", "helm", "sonar", "start", "pipeline-manager", "python-toolset", "node16-npm-toolset"]
+        image: ["package-image", "finish", "go-toolset", "gradle-toolset", "helm", "sonar", "start", "pipeline-manager", "python-toolset", "node16-npm-toolset", "node18-npm-toolset"]
     permissions:
       contents: read
       packages: write

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,10 @@ listed in the changelog.
 
 - Add new default npm toolset with Node.js 18 ([#585](https://github.com/opendevstack/ods-pipeline/issues/585))
 
+### Fixed
+
+- npm-toolset tests fail with new release of ubi8 Node.js image ([#650](https://github.com/opendevstack/ods-pipeline/issues/650))
+
 ## [0.8.0] - 2023-01-06
 
 ### Added

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,9 @@ listed in the changelog.
 
 ## [Unreleased]
 
+### Added
+
+- Add new default npm toolset with Node.js 18 ([#585](https://github.com/opendevstack/ods-pipeline/issues/585))
 
 ## [0.8.0] - 2023-01-06
 

--- a/Makefile
+++ b/Makefile
@@ -154,6 +154,7 @@ start-ods-builds: ## Start builds for each ODS BuildConfig
 	oc start-build ods-gradle-toolset
 	oc start-build ods-helm
 	oc start-build ods-node16-npm-toolset
+	oc start-build ods-node18-npm-toolset
 	oc start-build ods-pipeline-manager
 	oc start-build ods-python-toolset
 	oc start-build ods-sonar

--- a/build/package/Dockerfile.node16-npm-toolset
+++ b/build/package/Dockerfile.node16-npm-toolset
@@ -14,9 +14,9 @@ WORKDIR /app
 
 USER root
 
-RUN mkdir -p /.npm /.npm-global && \
-    chown -R 1001:0 /app /.npm /.npm-global $HOME && \
-    chmod -R g=u /app /.npm /.npm-global $HOME
+RUN mkdir -p /.npm $HOME/.npm-global/lib && \
+    chown -R 1001:0 /app /.npm $HOME && \
+    chmod -R g=u /app /.npm $HOME
 
 # Add scripts
 COPY build/package/scripts/build-npm.sh /usr/local/bin/build-npm

--- a/build/package/Dockerfile.node18-npm-toolset
+++ b/build/package/Dockerfile.node18-npm-toolset
@@ -1,0 +1,36 @@
+FROM registry.access.redhat.com/ubi8/nodejs-18:1
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+ENV NPM_CONFIG_PREFIX=$HOME/.npm-global \
+    LANG=en_US.UTF-8 \
+    LC_ALL=en_US.UTF-8
+
+RUN echo node version: $(node --version) && \
+    echo npm version: $(npm --version) && \
+    echo npx version: $(npx --version)
+
+WORKDIR /app
+
+USER root
+
+RUN mkdir -p /.npm /.npm-global && \
+    chown -R 1001:0 /app /.npm /.npm-global $HOME && \
+    chmod -R g=u /app /.npm /.npm-global $HOME
+
+# Add scripts
+COPY build/package/scripts/build-npm.sh /usr/local/bin/build-npm
+COPY build/package/scripts/cache-build.sh /usr/local/bin/cache-build
+COPY build/package/scripts/copy-build-if-cached.sh /usr/local/bin/copy-build-if-cached
+COPY build/package/scripts/copy-artifacts.sh /usr/local/bin/copy-artifacts
+COPY build/package/scripts/supply-sonar-project-properties-default.sh /usr/local/bin/supply-sonar-project-properties-default
+RUN chmod +x /usr/local/bin/build-npm && \
+    chmod +x /usr/local/bin/cache-build && \
+    chmod +x /usr/local/bin/copy-build-if-cached && \
+    chmod +x /usr/local/bin/copy-artifacts && \
+    chmod +x /usr/local/bin/supply-sonar-project-properties-default
+
+# Add sonar-project.properties
+COPY build/package/sonar-project.properties.d/npm.properties /usr/local/default-sonar-project.properties
+
+USER 1001

--- a/build/package/Dockerfile.node18-npm-toolset
+++ b/build/package/Dockerfile.node18-npm-toolset
@@ -14,9 +14,9 @@ WORKDIR /app
 
 USER root
 
-RUN mkdir -p /.npm /.npm-global && \
-    chown -R 1001:0 /app /.npm /.npm-global $HOME && \
-    chmod -R g=u /app /.npm /.npm-global $HOME
+RUN mkdir -p /.npm $HOME/.npm-global/lib && \
+    chown -R 1001:0 /app /.npm $HOME && \
+    chmod -R g=u /app /.npm $HOME
 
 # Add scripts
 COPY build/package/scripts/build-npm.sh /usr/local/bin/build-npm

--- a/deploy/ods-pipeline/charts/images/docker/Dockerfile.node18-npm-toolset
+++ b/deploy/ods-pipeline/charts/images/docker/Dockerfile.node18-npm-toolset
@@ -1,0 +1,10 @@
+ARG imageTag="latest"
+
+FROM ghcr.io/opendevstack/ods-pipeline/ods-node18-npm-toolset:$imageTag
+
+ARG privateCertServer
+USER root
+RUN if [ -n "${privateCertServer}" ]; then openssl s_client -showcerts -connect "${privateCertServer}" </dev/null \
+    | sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > "/etc/pki/ca-trust/source/anchors/${privateCertServer%:*}.pem" && \
+    update-ca-trust; fi
+USER 1001

--- a/deploy/ods-pipeline/charts/images/templates/bc-ods-node18-npm-toolset.yaml
+++ b/deploy/ods-pipeline/charts/images/templates/bc-ods-node18-npm-toolset.yaml
@@ -1,0 +1,30 @@
+{{if or .Values.global.enabledTasks.buildNPM .Values.node18NPMToolset}}
+kind: BuildConfig
+apiVersion: build.openshift.io/v1
+metadata:
+  name: ods-node18-npm-toolset
+  labels:
+    {{- include "chart.labels" . | nindent 4}}
+spec:
+  nodeSelector: null
+  output:
+    to:
+      kind: ImageStreamTag
+      name: 'ods-node18-npm-toolset:{{.Values.global.imageTag | default .Chart.AppVersion}}'
+  resources: {}
+  successfulBuildsHistoryLimit: 5
+  failedBuildsHistoryLimit: 5
+  postCommit: {}
+  strategy:
+    type: Docker
+    dockerStrategy:
+      buildArgs:
+        - name: imageTag
+          value: '{{.Values.global.imageTag | default .Chart.AppVersion}}'
+        - name: privateCertServer
+          value: '{{.Values.privateCertServer}}'
+  source:
+    dockerfile: |-
+      {{- .Files.Get "docker/Dockerfile.node18-npm-toolset" | nindent 6}}
+  runPolicy: Serial
+{{end}}

--- a/deploy/ods-pipeline/charts/images/templates/is-ods-node18-npm-toolset.yaml
+++ b/deploy/ods-pipeline/charts/images/templates/is-ods-node18-npm-toolset.yaml
@@ -1,0 +1,10 @@
+{{if or .Values.global.enabledTasks.buildNPM .Values.node18NPMToolset}}
+apiVersion: image.openshift.io/v1
+kind: ImageStream
+metadata:
+  name: ods-node18-npm-toolset
+  labels:
+    {{- include "chart.labels" . | nindent 4}}
+  annotations:
+    "helm.sh/resource-policy": keep
+{{end}}

--- a/deploy/ods-pipeline/charts/tasks/templates/task-ods-build-npm.yaml
+++ b/deploy/ods-pipeline/charts/tasks/templates/task-ods-build-npm.yaml
@@ -105,9 +105,9 @@ spec:
       type: string
       default: "false"
     - name: node-version
-      description: "Node.js version to use - supported versions: 16"
+      description: "Node.js version to use - supported versions: 16, 18"
       type: string
-      default: "16"
+      default: "18"
     - name: build-dir
       description: >-
         Must match the directory into which `npm run build` places files.

--- a/docs/architecture/component-namespaced-installation.puml
+++ b/docs/architecture/component-namespaced-installation.puml
@@ -28,7 +28,8 @@ Container_Boundary(c1, "Namespaced ODS Pipeline Installation"){
 
     Boundary(images, "Images") {
         Component(image_go_toolset, "ods/go-toolset", "Container Image", "Go, golangci-lint, build script")
-        Component(image_node16_npm_toolset, "ods/node16-npm-toolset", "Container Image", "Node.js, npm, build script")
+        Component(image_node16_npm_toolset, "ods/node16-npm-toolset", "Container Image", "Node.js 16, npm, build script")
+        Component(image_node18_npm_toolset, "ods/node18-npm-toolset", "Container Image", "Node.js 18, npm, build script")
         Component(image_python_toolset, "ods/python-toolset", "Container Image", "Python, build script")
         Component(image_gradle_toolset, "ods/gradle-toolset", "Container Image", "JDK, Gradle, build script")
         Component(image_buildah, "ods/buildah", "Container Image", "Buildah, Aqua scanner")
@@ -45,6 +46,7 @@ Rel(task_build_go, image_go_toolset, "uses")
 Rel(task_build_go, image_sonar, "uses")
 
 Rel(task_build_npm, image_node16_npm_toolset, "uses")
+Rel(task_build_npm, image_node18_npm_toolset, "uses")
 Rel(task_build_npm, image_sonar, "uses")
 
 Rel(task_build_python, image_python_toolset, "uses")

--- a/docs/design/software-design-specification.adoc
+++ b/docs/design/software-design-specification.adoc
@@ -303,7 +303,7 @@ Supplies default SonarQube project properties file if required (SDS-SHARED-3).
 | `ods-build-npm` Task resource
 a| The task defines two steps:
 
-. Build Node.js applications using npm (referencing SDS-TASK-17 and executing SDS-TASK-18).
+. Build Node.js applications using npm (referencing SDS-TASK-17 or SDS-TASK-30 and executing SDS-TASK-18).
   This step supports build skipping (executing SDS-SHARED-5 and/or SDS-SHARED-4 if enabled with parameter `cache-build`)
 . Analyze source code (referencing SDS-SHARED-1 and executing SDS-SHARED-2)
 
@@ -343,6 +343,10 @@ Supplies default SonarQube project properties file if required (SDS-SHARED-3).
 | SDS-TASK-28
 | `npm.properties` properties file
 | Default configuration for npm SonarQube project.
+
+| SDS-TASK-30
+| `ods-node18-npm-toolset` container image
+| Container image to build Node.js applications using npm. Based on `ubi8/nodejs-18` (SDS-EXT-32), includes SDS-SHARED-3, SDS-TASK-18 and SDS-TASK-28.
 |===
 
 ==== `ods-package-image` task
@@ -667,6 +671,12 @@ a| The script installs the Helm chart located in `deploy/ods-pipeline`. Further,
 | 0.36.0
 | Security scanner CLI.
 | https://www.trivy.dev/
+
+| SDS-EXT-32
+| NodeJS 18 for UBI 8
+| 1
+| Node.js 18 available as container is a base platform for building and running various Node.js 18 applications and frameworks. It is maintained by Red Hat and updated regularly.
+| https://catalog.redhat.com/software/containers/ubi8/nodejs-18/6278e5c078709f5277f26998
 
 |===
 

--- a/docs/tasks/ods-build-npm.adoc
+++ b/docs/tasks/ods-build-npm.adoc
@@ -103,8 +103,8 @@ without leading `./` and trailing `/`.
 
 
 | node-version
-| 16
-| Node.js version to use - supported versions: 16
+| 18
+| Node.js version to use - supported versions: 16, 18
 
 
 | build-dir

--- a/test/tasks/ods-build-npm_test.go
+++ b/test/tasks/ods-build-npm_test.go
@@ -168,5 +168,30 @@ func TestTaskODSBuildNPM(t *testing.T) {
 					)
 				},
 			},
+			"build javascript app using node16": {
+				WorkspaceDirMapping: map[string]string{"source": "javascript-sample-app"},
+				PreRunFunc: func(t *testing.T, ctxt *tasktesting.TaskRunContext) {
+					wsDir := ctxt.Workspaces["source"]
+					ctxt.ODS = tasktesting.SetupGitRepo(t, ctxt.Namespace, wsDir)
+					ctxt.Params = map[string]string{
+						"sonar-skip":   "true",
+						"node-version": "16",
+					}
+				},
+				WantRunSuccess: true,
+				PostRunFunc: func(t *testing.T, ctxt *tasktesting.TaskRunContext) {
+					wsDir := ctxt.Workspaces["source"]
+					checkFilesExist(t, wsDir,
+						filepath.Join(pipelinectxt.XUnitReportsPath, "report.xml"),
+						filepath.Join(pipelinectxt.CodeCoveragesPath, "clover.xml"),
+						filepath.Join(pipelinectxt.CodeCoveragesPath, "coverage-final.json"),
+						filepath.Join(pipelinectxt.CodeCoveragesPath, "lcov.info"),
+						filepath.Join(pipelinectxt.LintReportsPath, "report.txt"),
+						"docker/dist/src/index.js",
+						"docker/dist/package.json",
+						"docker/dist/package-lock.json",
+					)
+				},
+			},
 		})
 }

